src/native/s3_client.c

```diff
@@ -0,0 +1,389 @@
+            AWS_LS_S3_META_REQUEST,
+            "id=%p: invoking S3MetaRequest.onResponseHeaders callback. Code: %d",
+            (void *)meta_request,
+            response_status);
+
+        (*env)->CallVoidMethod(
+            env,
+            callback_data->java_s3_meta_request_response_handler_native_adapter,
+            s3_meta_request_response_handler_native_adapter_properties.onResponseHeaders,
+            response_status,
+            java_headers_array);
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onResponseHeaders callback",
+                (void *)meta_request);
+        }
+    }
+}
+
+static void s_on_s3_meta_request_finish_callback(
+    struct aws_s3_meta_request *meta_request,
+    const struct aws_s3_meta_request_result *meta_request_result,
+    void *user_data) {
+
+    (void)meta_request;
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        (struct s3_client_make_meta_request_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    if (callback_data->java_s3_meta_request_response_handler_native_adapter != NULL) {
+        (*env)->CallVoidMethod(
+            env,
+            callback_data->java_s3_meta_request_response_handler_native_adapter,
+            s3_meta_request_response_handler_native_adapter_properties.onFinished,
+            meta_request_result->error_code);
+
+        if (aws_jni_check_and_clear_exception(env)) {
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_META_REQUEST,
+                "id=%p: Ignored Exception from S3MetaRequest.onFinished callback",
+                (void *)meta_request);
+        }
+    }
+}
+
+JNIEXPORT jlong JNICALL Java_software_amazon_awssdk_crt_s3_S3Client_s3ClientMakeMetaRequest(
+    JNIEnv *env,
+    jclass jni_class,
+    jlong jni_s3_client,
+    jobject java_s3_meta_request_jobject,
+    jint meta_request_type,
+    jbyteArray jni_marshalled_message_data,
+    jobject jni_http_request_body_stream,
+    jobject java_response_handler_jobject) {
+    (void)jni_class;
+
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+    struct aws_s3_client *client = (struct aws_s3_client *)jni_s3_client;
+
+    struct s3_client_make_meta_request_callback_data *callback_data =
+        aws_mem_acquire(allocator, sizeof(struct s3_client_make_meta_request_callback_data));
+
+    AWS_FATAL_ASSERT(callback_data);
+
+    jint jvmresult = (*env)->GetJavaVM(env, &callback_data->jvm);
+    (void)jvmresult;
+    AWS_FATAL_ASSERT(jvmresult == 0);
+
+    callback_data->java_s3_meta_request = (*env)->NewGlobalRef(env, java_s3_meta_request_jobject);
+    callback_data->java_s3_meta_request_response_handler_native_adapter =
+        (*env)->NewGlobalRef(env, java_response_handler_jobject);
+
+    AWS_FATAL_ASSERT(callback_data->java_s3_meta_request != NULL);
```

Comment:
move this up to just after the creation.  In fact, delete line 304/5 entirely, per below

Reply:
Done